code/espurna/button.ino

```diff
@@ -45,12 +57,44 @@ unsigned char buttonCount() {
     return _buttons.size();
 }
 
+
+
+const uint8_t _buttonMapReleased(unsigned char index, uint8_t count, uint8_t length) {
+    return (
+        (1 == count) ? (
+            (length > _buttons[index].longLongClickDelay) ? BUTTON_EVENT_LNGLNGCLICK :
+            (length > _buttons[index].longClickDelay) ? BUTTON_EVENT_LNGCLICK : BUTTON_EVENT_CLICK
+        ) :
+        (2 == count) ? BUTTON_EVENT_DBLCLICK :
+        (3 == count) ? BUTTON_EVENT_TRIPLECLICK :
+        BUTTON_EVENT_NONE
+    );
+}
+
+const uint8_t _buttonMapEvent(unsigned char index, uint8_t event, uint8_t count, uint16_t length) {
+    return (
+        (event == EVENT_PRESSED) ? BUTTON_EVENT_PRESSED :
+        (event == EVENT_CHANGED) ? BUTTON_EVENT_CLICK :
+        (event == EVENT_RELEASED) ? _buttonMapReleased(index, count, length) :
+        BUTTON_EVENT_NONE
+    );
+}
+
 #if MQTT_SUPPORT
 
 void buttonMQTT(unsigned char id, uint8_t event) {
     char payload[4] = {0};
     itoa(event, payload, 10);
-    mqttSend(MQTT_TOPIC_BUTTON, id, payload, false, false); // 1st bool = force, 2nd = retain
+    mqttSend(MQTT_TOPIC_BUTTON, id, payload, false, getSetting({"btnMqttRetain", id})); // 1st bool = force, 2nd = retain
```

Comment:
Note that 2nd argument's type is always mapped to the return type, getSetting<...>() syntax can be avoided.
(preferably PREPROCESS_FLAGS are adjusted to be the correct type directly instead of casting)
getSetting(key) always returns String obj as-is from the underlying storage, since we don't know the type beforehand

Reply:
So `getSetting({"btnMqttRetain", id}, false)` is preferred over `getSetting<bool>({"btnMqttRetain", id})` ?

Reply:
Yes, since there is no `getSetting<bool>(key)`
Which could be implemented in theory, but instead of explicit `false` then you'd need to remember what exactly is the default of the type that you want.